src/adapter/src/coord/sequencer/inner/create_materialized_view.rs

```diff
@@ -673,29 +684,41 @@ impl Coordinator {
         .map(StageResult::Response)
     }
 
-    /// Select the initial `as_of` and `until` frontiers for a materialized view.
+    /// Select the initial `dataflow_as_of`, `storage_as_of`, and `until` frontiers for a
+    /// materialized view.
     fn select_timestamps(
         &self,
         id_bundle: CollectionIdBundle,
         refresh_schedule: Option<&RefreshSchedule>,
-    ) -> Result<(Antichain<mz_repr::Timestamp>, Antichain<mz_repr::Timestamp>), AdapterError> {
-        // Normally, `as_of` should be the least_valid_read.
-        let mut as_of = self.least_valid_read(&id_bundle);
-
-        // But for MVs with non-trivial REFRESH schedules, it's important to set the `as_of` to the
-        // first refresh. This is because we'd like queries on the MV to block until the first
-        // refresh (rather than to show an empty MV).
+    ) -> Result<
+        (
+            Antichain<mz_repr::Timestamp>,
+            Antichain<mz_repr::Timestamp>,
+            Antichain<mz_repr::Timestamp>,
+        ),
+        AdapterError,
+    > {
+        // The dataflow's `as_of` should always be simply `least_valid_read`.
+        let dataflow_as_of = self.least_valid_read(&id_bundle);
```

Comment:
Would it make sense to choose `greatest_valid_read` instead for REFRESH MVs with a refresh time in the future? If the difference between `least_valid_read` and `greatest_valid_read` is large, choosing the latter would avoid crunching through a lot of times needlessly, which would reduce the hydration memory spike.

Reply:
Good idea! Done. But please carefully check my changes, because now the code became non-trivial.